lib/src/utils/screen_loading/screen_loading_manager.dart

```diff
@@ -0,0 +1,220 @@
+    final microTimeStamp = DateTime.now().microsecondsSinceEpoch;
+    final uiTraceId = DateTime.now().millisecondsSinceEpoch;
+    APM.startCpUiTrace(screenName, microTimeStamp, uiTraceId);
+    _currentUiTrace = UiTrace(screenName, traceId: uiTraceId);
+  }
+
+  /// @nodoc
+  @internal
+  Future<bool> startScreenLoadingTrace(
+    String screenName, {
+    required int startTimeInMicroseconds,
+  }) async {
+    final isScreenLoadingMonitoringEnabled =
+        await FlagsConfig.ScreenLoading.isEnabled();
+    if (!isScreenLoadingMonitoringEnabled) {
+      log("Unable to start Screen loading capture, as ${FlagsConfig.ScreenLoading.name} feature is disabled.",
+          name: APM.tag);
+      return false;
+    }
+    // only accepts the first widget with the same name
+    if (screenName == _currentUiTrace?.screenName &&
+        _currentUiTrace?.didStartScreenLoading == false) {
+      final trace = ScreenLoadingTrace(
+        screenName,
+        startTimeInMicroseconds: startTimeInMicroseconds,
+      );
+      debugPrint('${APM.tag} starting screen loading trace — screenName: $screenName, startTimeInMicroseconds: $startTimeInMicroseconds');
+      _currentUiTrace?.didStartScreenLoading = true;
+      _currentScreenLoadingTrace = trace;
+      return true;
+    }
+    debugPrint('${APM.tag} failed to start screen loading trace — screenName: $screenName, startTimeInMicroseconds: $startTimeInMicroseconds');
+    debugPrint('${APM.tag} didStartScreenLoading: ${_currentUiTrace?.didStartScreenLoading}, isSameName: ${screenName ==
+        _currentUiTrace?.screenName}');
+    return false;
+  }
+
+  /// @nodoc
+  @internal
+  Future<bool> reportScreenLoading(
+    ScreenLoadingTrace? trace, {
+    required int endTimeInMicroseconds,
+  }) async {
+    int? duration;
+    final isScreenLoadingMonitoringEnabled =
+        await FlagsConfig.ScreenLoading.isEnabled();
+    if (!isScreenLoadingMonitoringEnabled) {
+      log('Unable to report Screen loading capture, as ${FlagsConfig.ScreenLoading.name} feature is disabled.',
+          name: APM.tag);
+    }
+
+    // Handles if [endScreenLoading] was called prematurely
+    // As the endScreenLoading only extends the current screen loading trace
+    if (prematurelyEndedTraces.contains(trace)) {
+      log(
+        'Ending screen loading prematurely, as Screen loading was ended before full capture.',
+        name: APM.tag,
+      );
+      duration = 0;
+    }
+
+    // Only report the first screen loading trace with the same name as the active UiTrace
+    if (trace?.screenName == _currentScreenLoadingTrace?.screenName &&
+        _currentUiTrace?.didReportScreenLoading == false) {
+      trace!.duration = (trace.endTimeInMicroseconds ?? 0) - (trace.startTimeInMicroseconds ?? 0);
+      _currentUiTrace?.didReportScreenLoading = true;
+
+      APM.reportScreenLoading(
+        trace.startTimeInMicroseconds,
+        duration ?? trace.duration!,
+        _currentUiTrace?.traceId ?? 0,
+      );
```

Comment:
We are using the `duration` variable in line 125 to change the duration to `0` if the trace has been prematurely ended (depending on `prematurelyEndedTraces.contains(trace)`) to then use it with the `??` operator on line `136` when we are reporting the screen loading time.

I think this is a bit confusing because the places where it's defined, assigned a value, and then used in a condition are a bit separated from one another by other pieces of code.

I'd rather recommend defining a variable `isPrematurelyEnded` and adding an `if` condition or a ternary statement on the duration to report (something like `isPrematuralyEnded ? 0 : trace.duration`) as it states clearly that this part depends on `isPrematuralyEnded`.

Another note as well, I feel like the idea that prematurely-ended traces get reported with duration `0` is very unnecessary and counterintuitive and will cause a lot of confusion for users even with proper documentation or API name. I think just reporting the time the user called `endScreenLoading` even if it's before the base screen loading time has ended is the better solution here. I'd love to hear your opinion on this and I'd also want to state that we don't need to match the native SDK's behavior in every aspect, especially when that behavior feels unintuitive for users.

Reply:
This flow was invalid. When `endScreenLoading` gets called before the capturing and report is done, we report the span with 0 value (the extended span). Afterwards when the capturing ends it gets called with the normal end duration. In case the `endScreenLoading` gets called again normally it overwrites the 0 value.

I've applied those changes where `reportScreenLoading` doesn't need to check for `prematurelyEndedTraces`.
The `endScreenLoading` checks for prematurely ended screen loading traces, logs it and sends duration as 0, and allows overwriting the value later on.